docs/proposal/20230317-irsa-support-for-awscluster.md

```diff
@@ -0,0 +1,131 @@
+
+The pieces to IRSA are easily created with the existing access for CAPA. By adding `AssociateOIDCProvider` to `AWSCluster` we can kick off a reconciliation process to generate all pieces necessary to utilize IRSA in your self-managed cluster. 
+
+### Goals
+
+1. On cluster create add all components to self-managed clusters to use IAM Roles for Service Accounts.
+2. On cluster delete remove all external dependencies from the AWS account.
+
+### Non-Goals/Future Work
+Migrate all IAM work for Managed cluster to the IAM service. 
+
+## Proposal
+- Create a boolean on `AWSCluster` called `AssociateOIDCProvider` to match the `AWSManagedCluster` API and have a default value of false.
+- Migrate the status types for `OIDCProvider` out of the experimental EKS APIs and into the v1beta2 APIs. 
+- Build an IAM cloud service and add a reconciler to work to persist all components required for IRSA, the logic is as follows.
+  1. Create a self-signed issuer for the workload cluster namespace to be used to make the pod identity webhook serving cert.
+  2. Generate the patch file and update kubeadm configs to write the patch to disk for the control plane nodes.
+  3. Create the Identity Provider in IAM pointed to the S3 bucket.
+  4. Pause the reconciler until the workload cluster online as we have created all the pieces we can without a working kube api, the AWSMachine controller has additional code to annotate the `AWSCluster` if a control plane node is up and if the management cluster has a kubeconfig which will unpause our reconciler.
+  5. Copy the JWKS and OpenID Configuration from the kubeapi to the S3 bucket.
+  6. Create all kube components in the workload cluster to run the pod-identity-webhook
+  7. Create the trust policy boilerplate configmap in the workload cluster 
+
+Identical to the EKS implementation, there will have a trust policy document boilerplate which references the ARN for the Identity Provider created in step 3. This can be used to generate IAM roles and the ARNs for those roles can be annotated on a service account. The pod-identity-webhook works watching by watching all service accounts and pods. When it finds a pod using a service account with the annotation it will inject AWS STS Tokens via environment variables generated from the role ARN. 
```

Comment:
```suggestion
Identical to the EKS implementation, a trust policy document boilerplate will reference the ARN for the Identity Provider created in step 3. This can be used to generate IAM roles, and the ARNs for those roles can be annotated on a service account. The pod-identity-webhook works by watching all service accounts and pods. When it finds a pod using a service account with the annotation, it will inject AWS STS Tokens via environment variables generated from the role ARN. 
```

>it will inject AWS STS Tokens via environment variables generated from the role ARN. 

Inject it into the POD?